coveo-functools/coveo_functools/flex.py

```diff
@@ -0,0 +1,300 @@
+
+    return value  # type: ignore[no-any-return]
+
+
+@overload
+def _generate_wrapper(obj: RealClass) -> WrappedClass:
+    ...
+
+
+@overload
+def _generate_wrapper(obj: RealFunction) -> WrappedFunction:
+    ...
+
+
+def _generate_wrapper(obj: RealObject) -> WrappedObject:
+    """Generates a wrapper over obj."""
+    if inspect.isclass(obj):
+        return _generate_class_wrapper(obj)  # type: ignore[arg-type]
+
+    return _generate_callable_wrapper(obj)
+
+
+def _generate_callable_wrapper(fn: RealFunction) -> WrappedFunction:
+    """Class decorators"""
+
+    @functools.wraps(fn)
+    def wrapper(*args: Any, **kwargs: Any) -> T:
+        value: T = fn(*args, **_remap_and_convert(fn, kwargs))
+        if hasattr(value, "__dict__"):
+            value.__dict__[RAW_KEY] = kwargs
+        return value
+
+    return wrapper
+
+
+def _generate_class_wrapper(obj: RealClass) -> WrappedClass:
+    """Function decorators"""
+    fn: RealFunction = obj.__init__
+
+    @functools.wraps(fn)
+    def new_init(*args: Any, **kwargs: Any) -> None:
+        setattr(args[0], RAW_KEY, kwargs)  # set the raw data on self
+        fn(*args, **_remap_and_convert(fn, kwargs))
+
+    obj.__init__ = new_init
+    return obj
+
+
+def _remap_and_convert(fn: RealFunction, dirty_kwargs: Dict[str, Any]) -> Dict[str, Any]:
+    """Orchestrates the deserialization of `dirty_kwargs` into `fn`."""
+    mapped_kwargs = unflex(fn, dirty_kwargs)
+    converted_kwargs = {}
+
+    for arg_name, arg_type in find_annotations(fn).items():
+        if arg_name not in mapped_kwargs:
+            continue  # this may be ok, for instance if the target argument has a default
+        converted_kwargs[arg_name] = deserialize(mapped_kwargs[arg_name], hint=arg_type)
+
+    return converted_kwargs
+
+
+def _resolve_hint(thing: TypeHint) -> Tuple[TypeHint, Tuple[TypeHint, ...]]:
+    """
+    Transform e.g. List[Union[str, bool]] into (list, Union[str, bool]).
+    Also validates that the annotation is supported and removes "NoneType" if present.
+    """
+    origin = get_origin(thing) or thing
+    args = {*get_args(thing)}
+
+    # typing implementation detail -- At runtime, Optional exists as Union[None, ...]
+    assert origin is not Optional
+
+    if origin is dict:
+        # In this case, args isn't a collection of types; rather a (key_type, value_type).
+        # Dict[str, Any] would therefore come out as a (Dict, Union[str, Any]) which is wrong.
+        # since we don't do anything special for dicts anyway yet, drop the arg types and carry on.
+        return dict, ()
+
+    if not args.difference(PASSTHROUGH_TYPES):
+        # If all containing types are passthrough types, everything shall be fine.
+        return origin, tuple(args)
+
+    # Remove NoneType if it's present. If the value is given as None, we return None, no questions asked,
+    # so we really don't need to keep this information.
+    args.discard(type(None))

```

Comment:
If I understand correctly, you should call this line _before_ you 

```
    if not args.difference(PASSTHROUGH_TYPES):
        # If all containing types are passthrough types, everything shall be fine.
        return origin, tuple(args)
```

(that is, above line `262`) because the `NoneType` is in the `PASSTHROUGH_TYPES`.

Reply:
The functional reason for stripping out NoneType, is to transform e.g. `Optional[str]`, which comes in as `Union[str, None]`, which I really want to read as `Union[str]` or just str really...

Your suggestion wouldn't change the expected behavior. It's ok if NoneType is compared against PassThrough types, for instance `Optional[Union[str, bool]]` would be OK regardless if I compare (str, bool, NoneType) or (str, bool) vs passthrough.

Reply:
for the record: I ended up applying this suggestion since the function's comment mentions that None is removed from the return value, which wasn't true in the case of the passthrough.